EZP-26098: Add generic command to recreate search engine index

alongosz opened the pull request:
> Status: **Ready for a review**.
> Implements: [EZP-26098](https://jira.ez.no/browse/EZP-26098).

This PR adds generic `ezplatform:reindex` command able to recreate Search Engine Index for any Search Engine extending Search Engine Indexer Service.
The Service should be tagged as `ezpublish.searchEngineIndexer` with an alias corresponding to a specific search engine (legacy, solr, elasticsearch).

**TODO**:
- [x] Add abstract Indexer Service, to be extended by each Search Engine  (af09e66).
- [x] Add `ezplatform:reindex` command to `EzPublishCoreBundle` (2ef2440).
- [x] Add Indexer Service for Legacy Search Engine (d4ff5c4).
- [x] Add Indexer Service for Elasticsearch Search Engine (b3b31f0).
- [x] Fix Elasticsearch Handler service definition (6c4a331).
- [x] Add Indexer Service for Solr ([Solr PR #68](https://github.com/ezsystems/ezplatform-solr-search-engine/pull/68)).

_Note: This PR was rebased, removing any code related to previous approach. Old code is available [here](https://github.com/ezsystems/ezpublish-kernel/compare/master...alongosz:ezp-26098-backup-20160926)._

joaoinacio commented on eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php:
```diff
@@ -0,0 +1,224 @@
+                    break;
+                }
+
+                $contentObjects[] = $this->persistenceHandler->contentHandler()->load(
```
IIRC, this can throw an exception (not found). If so, shouldn't it be handled/logged so that it doesn't break further indexing?
(see also https://github.com/ezsystems/ezplatform-solr-search-engine/commit/342d224141c5649774d181a2e5e9a091f3268fe9 )

alongosz replied:
Yes :) I've already added this to a *TODO* yesterday and fixed this in eee88f4 today :)

joaoinacio commented on eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php:
```diff
@@ -0,0 +1,224 @@
+                    break;
+                }
+
+                $locations[] = $this->persistenceHandler->locationHandler()->load($row['node_id']);
```
same as above

alongosz replied:
Added exception handling in eee88f4 to keep code consistent, but haven't seen a case when an error occurs here.

alongosz commented on eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php:
```diff
@@ -0,0 +1,254 @@
+        $this->output->writeln('Indexing Content...');
+
+        /** @var \Symfony\Component\Console\Helper\ProgressBar $progress */
+        $progress = new ProgressBar($this->output);
```
Here I decided to use `ProgressBar` instead of [deprecated `ProgressHelper`](https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Console/Helper/ProgressHelper.php#L25). In the other commands it is retrieved using `$this->getHelperSet()->get('progress')`. Right now I don't know how to use it for `ProgressBar`.

Using `ProgressBar` also allows us to get rid of [such ugly workarounds when logging other messages](https://github.com/alongosz/ezpublish-kernel/blob/40f005e1f3e724f0c3af62ae4e412a4e033f5691/eZ/Bundle/EzPublishLegacySearchEngineBundle/Command/CreateIndexCommand.php#L176-L177)

joaoinacio replied:
Looking at the [SF docs](http://symfony.com/doc/current/components/console/helpers/progressbar.html) creating a new instance seems to be the correct / only usage (which makes sense, since each one will be different)

alongosz commented on eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php:
```diff
@@ -0,0 +1,254 @@
+
+            foreach ($contentObjects as $contentObject) {
+                try {
+                    $this->searchHandler->indexContent($contentObject);
```
Here there is a change of the way how we bulk-index data. Instead of [exposing documents](https://github.com/ezsystems/ezplatform-solr-search-engine/blob/1.0/bundle/Command/SolrCreateIndexCommand.php#L106) to the command I sort of moved this loop from Handler to the Command. The reason for this is that not every search engine has a concept of a document.

@pspanja: Even [`bulkIndexContent` in Solr v1.0.2.2](https://github.com/ezsystems/ezplatform-solr-search-engine/blob/v1.0.2.2/lib/Handler.php#L244-L255) has just a simple loop without any search engine specific stuff, that's why I think moving it here is not such a bad idea :)

This change allows us to handle exceptions that may occur for a single content - without disregarding objects in a bulk.

andrerom replied:
> @pspanja: Even bulkIndexContent in Solr v1.0.2.2 has just a simple loop without any search 
engine specific stuff, that's why I think moving it here is not such a bad idea :)

the difference here is that this code indexes one object at the time, the other one was indexing several in batches, ref `$this->gateway->bulkIndexDocuments($documents);`.

alongosz replied:
I looked into how exactly Solr bulk-indexes content and now I see that I shouldn't skip this.
I can see two possible solutions for this:

1. Go back to using [`bulkIndexContent`](https://github.com/ezsystems/ezplatform-solr-search-engine/blob/master/lib/Handler.php#L244-L259) which silences exceptions occuring while mapping documents. I didn't see any of those exceptions in current versions of our demo applications, so it's possible there was just a bug which is now fixed. 
(We could also pass `LoggerInterface` to `bulkIndexContent` but AFAIR we weren't happy with such solutions).
2. To be able to handle in the command exceptions that may occur while bulk-indexing we could introduce in search handlers a method `prepareContentForBulkIndexing` which creates an array of some new `ValueObject` containing data for `bulkIndexContent`. Exceptions would be caught while creating this array. However this seems to me like overcomplicating things, but maybe you'll like it.

Of course I'm open to other suggestions.

Ping @andrerom @pspanja

andrerom replied:
I would go for 1, its cleaner then callbacks, however question is how this behaves with ProgressBar?

2. was kind of dismissed by Petar, when we considered exposing `generateDocument(Content $content) : Document`, and `bulkIndexDocuments(Document[] $documents)`, as that approach would make sure indexing command can catch load errors, and also optimize load as api matures.

andrerom replied:
In regards to 1, as you said it will cause issues for progress bar, what we can do is actually do an internal implement of LoggerInterface that takes LoggerInterface and ProgressBar as input and handles the logic of clearing progress bar on any warnings etc.

So these are the two options we have then.

alongosz replied:
>  question is how this behaves with ProgressBar

Good point, missed that issue. ProgressBar also needs to be passed in this case because clean logging while ProgressBar is runing looks like this:
```
$progress->clear();
$logger->warning($message);
$progress->display();
```

andrerom replied:
> Good point, missed that issue. ProgressBar also needs to be passed in this case because 

sure, but handler should not have to know about this, so if we go this route we need a internal implementation of `LoggerInterface` as explained above.

alongosz commented on eZ/Publish/SPI/Search/Indexing.php:
```diff
@@ -0,0 +1,26 @@
+     * Make indexing operation permanent if Search Handler needs and supports it.
+     * Note: it is a responsibility of a Search Handler to decide if this is needed and how to do this.
+     */
+    public function commitIndex();
```
The reason I introduced here the new method instead of reusing existing in Solr and LSE `commit` method is BC. `commit` has one optional parameter `$flush` which for generic purpose makes no sense. Solr Search Handler itself when implementing `commitIndex` should decide if this should be hard or soft commit. LSE implementation of this method is empty.

andrerom replied:
I would rather go with resuing `commit` but we can indeeed skip the flush argument:
```php
    /**
     * Commits changes to search index, making it available for search.
     *
     *  Search engines implementing this should expose configuration to specify if this should
     * do a flush, aka hard commit, meaning changes should be forced to be written to durable
     * storage or not. By default this should be the case, but if transaction logs are enabled
     * it can be safe to optimize performance and allow disabling flushing.
     * 
     * @param bool $forceFlush
     */
    public function commit();
```

alongosz commented:
@andrerom: this is ready for a review and comparison with #1740 approach.

andrerom commented:
So on this there are a few issues:
- content index will be done first and locations second
- _minor, can be done later_ bulk index is skipped
- new commitIndex introduced, while existing one could have been re used instead. Indexing command could expose flag for flushing index right away for instance, these are generic concepts across solr and elasticsearch people kind of need to know about anyway.

alongosz commented:
> content index will be done first and locations second

Ok, I fixed this in 4da2b65 (relevant part starts [here](https://github.com/ezsystems/ezpublish-kernel/commit/4da2b65ea2badc15e9d963bc7ed32e38a834406b#diff-81257ebfd63be05261b508add93d35edR150)) - locations are now fetched during content fetching if main location exists and a search handler needs to index locations separately, like in the case of *Elasticsearch*.

alongosz commented:
We decided to omit bulk indexing for now. The `bulk_count` command argument was replaced by 
`--iteration-count` option as "bulk" might be misleading (by default `5`). Moreover, after each iteration, a commit is performed, unless disabled by `--no-iteration-commit` option.
Changed in bfa4ba8.

Review ping @andrerom @pspanja

andrerom commented on eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php:
```diff
@@ -0,0 +1,255 @@
+        $this->databaseHandler = $this->getContainer()->get('ezpublish.connection');
+        if (!$this->searchHandler instanceof Indexing) {
+            throw new RuntimeException(
+                'Expected to find Search Engine Handler implementing Indexing but found something else.'
```
`Expected to find Search Engine Handler implementing "eZ\Publish\SPI\Search\Indexing" to be able to re index, but currently configured handler "${handlerClassName}" does not.`

andrerom replied:
Also, this checks `Indexing`, however without `ContentIndexing` it will end up deleting index but not re create it. This should either not be allowed, or be warned and prompted to user to acknowledge _(unless a force-yes option is set)_ that he agrees to wipe out index.

alongosz replied:
Ok. Since we now index Locations while indexing content, I moved relevant code [here](https://github.com/alongosz/ezpublish-kernel/blob/0250f1843ca49783de549218f6ac2af186f79a47/eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php#L102-L105).

andrerom commented on eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php:
```diff
@@ -0,0 +1,255 @@
+        $this
+            ->setName('ezplatform:reindex')
+            ->setDescription('Recreate search engine index')
+            ->addOption('iteration-count', 'c', InputOption::VALUE_OPTIONAL, 'Number of objects to be indexed in a single iteration', 5)
```
can increase default to 20

andrerom commented on eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php:
```diff
@@ -0,0 +1,255 @@
+            ->setName('ezplatform:reindex')
+            ->setDescription('Recreate search engine index')
+            ->addOption('iteration-count', 'c', InputOption::VALUE_OPTIONAL, 'Number of objects to be indexed in a single iteration', 5)
+            ->addOption('no-iteration-commit', null, InputOption::VALUE_NONE, 'Do not commit after each iteration')
```
let's change this to `no-commit`, at first looks like it implies no iteration in current form.

andrerom commented on eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php:
```diff
@@ -0,0 +1,255 @@
+        $progress = new ProgressBar($this->output);
+        $progress->start($totalCount);
+        $i = 0;
+        $indexLocations = $this->searchHandler instanceof LocationIndexing;
```
move this inline as it is only used once, will increase readability.

andrerom commented on eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php:
```diff
@@ -0,0 +1,255 @@
+     * @param int $contentObjectId
+     * @return array Location nodes Ids
+     */
+    private function getContentObjectLocationNodeIds($contentObjectId)
```
`getContentLocationIds` you don't need to mic in legacy terms :)

andrerom commented on eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php:
```diff
@@ -0,0 +1,255 @@
+     * @param array $fields Select fields
+     * @return \PDOStatement
+     */
+    private function getContentObjectStmt(array $fields)
```
`getContentDbFieldsStmt` or something like it

alongosz commented:
Fixed all issues.
Looks like this is heading to a finish, so I rebased this PR and rewritten commit history (I was unhappy with scope & messages of my commits). Hope it is not that inconvenient at this point, but just in case I made a [backup of this PR](https://github.com/ezsystems/ezpublish-kernel/compare/master...alongosz:backup-pr-1738).

andrerom commented on eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php:
```diff
@@ -0,0 +1,252 @@
+                    $this->logWarning($progress, "Could not load current version of Content with id ${row['id']}, so skipped for indexing. Full exception: " . $e->getMessage());
+                }
+
+                if (!($this->searchHandler instanceof LocationIndexing) || empty($row['node_id'])) {
```
I see now you can greatly simplify `getContentDbFieldsStmt` and instead check for existing location using:
if (!($this->searchHandler instanceof LocationIndexing) || empty($content->getVersionInfo()->getContentInfo()->mainLocationId)) {

alongosz replied:
> I see now you can greatly simplify `getContentDbFieldsStmt` and instead check for existing location using:
> ```php
> // Skip location indexing if search engine does not use it, or if content does not have locations
> if (!($this->searchHandler instanceof LocationIndexing) || empty($content->getVersionInfo()->getContentInfo()->mainLocationId)) {
> ```

Nice one :) However my `$content` is a `\eZ\Publish\SPI\Persistence\Content` `ValueObject` in this context, so I accessed main `LocationId` via `$content->versionInfo->contentInfo->mainLocationId`. Also needed to reorganize code inside iteration loop. 
Changed in 02563ce.

pspanja commented:
@alongosz 

> Note: the command uses default siteaccess.
> To create index for non-default siteaccess use --siteaccess= option. 

What does that mean?
Index should be per-repository and not per-siteaccess.

andrerom commented on eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php:
```diff
@@ -0,0 +1,242 @@
+                    $contentObjects[] = $content;
+
+                    // Skip location indexing if search engine does not use it, or if content does not have locations
+                    if (!($this->searchHandler instanceof LocationIndexing) || empty($content->versionInfo->contentInfo->mainLocationId)) {
```
this can be moved to after line 170, adjusting catch block there to do a continue on error, so we only load locations that we managed to index content for. As a bonus this reduces nesting of the code which got worse now.

pspanja replied:
Since this method is called `createContentIndex()`, I would expect separate `createLocationIndex()`.
So maybe introduce that and call it from `execute()`, unless this was done on purpose, to have Locations indexed progressively with Content?

alongosz replied:
Indeed we're fetching Locations while indexing Content on purpose. 
Locations are indexed if `mainLocationId` of a `Content` exists. 
So, if I'm to move this part after line 170 it would have to be [inside this loop](https://github.com/alongosz/ezpublish-kernel/blob/02563ce994088324df8b6901863a3020d7e17886/eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php#L165-L171) which also causes too much nesting. 

Could also store all `mainLocationId`s in an array and fetch Locations after the line 170. 

I made a backup of this command with separate indexing of Locations [here](https://github.com/alongosz/ezpublish-kernel/blob/96f5878c2e304703dfe7c667eb6c483285309ed6/eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php#L89) to compare.

andrerom replied:
> So, if I'm to move this part after line 170 it would have to be inside this loop which also causes too much nesting.


well it won't need to be inside the try/catch as you do here, so one level less, and it will be skipped if content can not be indexed.

alongosz replied:
Ok, moved this as described above and extracted to a private method in c74483d.

alongosz commented:
> > Note: the command uses default siteaccess.
> > To create index for non-default siteaccess use --siteaccess= option. 

> What does that mean?
> Index should be per-repository and not per-siteaccess.

@pspanja: I meant that if you want to create search index for a different repository, you need to refer to siteaccess which is using that repository. There's no direct way to pass repository name AFAIK.

pspanja commented:
This new approach looks good to me :)

alongosz commented:
I'm ready to move forward with this PR. Due to many changes in the meantime I had to rebase it.

My main problem was testing it using Elasticsearch. I know that we don't support it as it's experimental, but this seemed to me like something just broken (despite working integration tests).

In 6c4a331 I've proposed a fix for Elasticsearch service definition and factory. I'm not sure why, but long time ago, in afb04c4, there were introduced changes breaking `ElasticsearchEngineFactory`. Moreover fixing this did not solve a problem. Using builtin `ezplatform:elasticsearch_create_index` command on clean installation of eZ Platform caused PHP to crash (core dumped) due to endless recursion inside Symfony Container for ES Handler. 

I've fixed this modifying service definition in proper Extension and Factory taking Solr as an example. I'm not sure if I did it right - but at least it works ;)

(Btw. 6c4a331 should probably be a separate PR backported to `6.6` but I want to get your opinion on the fix itself first).

Review ping @pspanja @andrerom

andrerom commented:
@pspanja Can you have a look at the ES thingy?

pspanja commented:
Change to ES factory looks OK. No backport is needed as it's not released yet, for now we mostly want to run integration tests with it.

pspanja commented on eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php:
```diff
@@ -0,0 +1,73 @@
+        $noCommit = $input->getOption('no-commit');
+
+        if (!is_numeric($iterationCount) || (int)$iterationCount < 1) {
+            throw new RuntimeException("'bulk_count' argument should be > 0, got '{$iterationCount}'");
```
`iteration-count` instead of `bulk_count`

alongosz replied:
Good catch, it's actually an option now, so I adjusted message in 2a52d07.

alongosz commented:
There's one issue raised by @andrerom in corresponding Solr PR which needs to be addressed here:
> ```yaml
>     tags:
>        - {name: ezpublish.searchEngineIndexer, alias: solr}
> ```
> Missed this, but we should not use ezpublish in new tags we make right? /cc @bdunogier

Should it be `ezplatform` then? // cc @pspanja @bdunogier

pspanja commented:
> Should it be ezplatform then?

Up to @bdunogier :)
But please update this for `$output->writeln(...)`, as already done on the Solr Indexer PR.

andrerom commented:
so we stay with the tag name to be consistent, so can merge as soon as `$output->writeln(...)` is fixed.
